[GPU Process] [FormControls] Add a ControlPart for Button

shallawa opened the pull request:
#### 13bd3d0bfff1b515cfc56ab9a9f9f249891f5831
<pre>
[GPU Process] [FormControls] Add a ControlPart for Button
<a href="https://bugs.webkit.org/show_bug.cgi?id=249878">https://bugs.webkit.org/show_bug.cgi?id=249878</a>
rdar://103692353

Reviewed by Aditya Keerthi.

ButtonPart will handle drawing the button form control. This covers the parts&apos;
types: Button, PushButton, DefaultButton and SquareButton. The following HTML
markup shows different buttons in the page:

&lt;button&gt;
&lt;input type=&quot;button&quot;&gt;
&lt;input type=&quot;submit&quot;&gt;
&lt;input type=&quot;reset&quot;&gt;
&lt;input type=&quot;file&quot;&gt;

On macOS, AppKit will be used to draw the platform controls through ButtonMac.

* Source/WebCore/Headers.cmake:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/controls/ButtonPart.h: Copied from Source/WebCore/platform/graphics/mac/controls/TextFieldMac.h.
* Source/WebCore/platform/graphics/controls/ControlFactory.h:
* Source/WebCore/platform/graphics/controls/PlatformControl.h:
(WebCore::PlatformControl::rectForBounds const):
* Source/WebCore/platform/graphics/mac/controls/ButtonMac.h: Copied from Source/WebCore/platform/graphics/mac/controls/TextFieldMac.h.
* Source/WebCore/platform/graphics/mac/controls/ButtonMac.mm: Added.
(WebCore::ButtonMac::ButtonMac):
(WebCore::ButtonMac::cellSize const):
(WebCore::ButtonMac::cellOutsets const):
(WebCore::ButtonMac::bezelStyle const):
(WebCore::ButtonMac::updateCellStates):
(WebCore::ButtonMac::rectForBounds const):
(WebCore::ButtonMac::draw):
* Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.h:
* Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.mm:
(WebCore::createButtonCell):
(WebCore::ControlFactoryMac::buttonCell const):
(WebCore::ControlFactoryMac::defaultButtonCell const):
(WebCore::ControlFactoryMac::createPlatformButton):
* Source/WebCore/platform/graphics/mac/controls/ControlMac.h:
* Source/WebCore/platform/graphics/mac/controls/ControlMac.mm:
(WebCore::ControlMac::setFocusRingClipRect):
(WebCore::ControlMac::inflatedRect):
* Source/WebCore/platform/graphics/mac/controls/TextFieldMac.h:
* Source/WebCore/platform/mac/ThemeMac.mm:
(WebCore::button):
(WebCore::ThemeMac::paint):
(WebCore::paintButton): Deleted.
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::createControlPart const):
* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::RenderThemeMac::canPaint const):
(WebCore::RenderThemeMac::canCreateControlPartForRenderer const):
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder&lt;ControlPart&gt;::decode):

Canonical link: <a href="https://commits.webkit.org/258492@main">https://commits.webkit.org/258492@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e2b466ccd143c8406c37f77ca34baa134fe9ea4e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/102095 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/11240 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/35164 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/111420 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/171596 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/106076 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/12214 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/2151 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/94476 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/109156 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/107875 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/9348 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/92623 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/37161 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/91226 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/24101 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/78886 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/4797 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/25528 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/4906 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/1970 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/10963 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/45024 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5824 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/6654 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/platform/mac/ThemeMac.mm:
```diff
@@ -493,61 +493,6 @@ static void setUpButtonCell(NSButtonCell *cell, ControlPartType type, const Cont
-
-    bool needsRepaint = ThemeMac::drawCellOrFocusRingWithViewIntoContext(buttonCell, context, inflatedRect, view, true, states.contains(ControlStates::States::Focused), deviceScaleFactor);
-    if (states.contains(ControlStates::States::Default))
-        [window setDefaultButtonCell:buttonCell];
```
It doesn't look like the call to `setDefaultButtonCell` is preserved anywhere.

shallawa replied:
I think the existing code is incorrect. It is supposed to set the default button cell before the drawing and restore it after the drawing. This code used to look like this (See https://bugs.webkit.org/show_bug.cgi?id=21775)

```
    NSView *view = scrollView->documentView();
    NSWindow *window = [view window];
    NSButtonCell *previousDefaultButtonCell = [window defaultButtonCell];

    if ((states & DefaultState) && [window isKeyWindow]) {
        [window setDefaultButtonCell:buttonCell];
        wkAdvanceDefaultButtonPulseAnimation(buttonCell);
    } else if ([previousDefaultButtonCell isEqual:buttonCell])
        [window setDefaultButtonCell:nil];

    [buttonCell drawWithFrame:NSRect(inflatedRect) inView:view];
    [buttonCell setControlView:nil];

    if (![previousDefaultButtonCell isEqual:buttonCell])
        [window setDefaultButtonCell:previousDefaultButtonCell];
```
But I think this order was changed mistakenly in https://trac.webkit.org/changeset/187961/webkit. This change made the drawing happens before setting up and restoring the default button.

shallawa replied:
I restored the order as it was before https://trac.webkit.org/changeset/187961/webkit.

pxlcoder commented on Source/WebCore/platform/graphics/mac/controls/ButtonMac.mm:
```diff
@@ -0,0 +1,151 @@
+    // these margins extend well past the boundaries of a native button cell's shadows.
+    static const IntOutsets cellOutsets[] = {
+        // top right bottom left
+        { 4, 6, 7, 6 },
```
Is there a plan to remove these values from `ThemeMac`?

shallawa replied:
Yes they will be removed once all the references to buttonSizes() and buttonMargins() are removed. I will fix this after refactoring the painting code for all the controls.

shallawa committed:
[GPU Process] [FormControls] Add a ControlPart for Button
https://bugs.webkit.org/show_bug.cgi?id=249878
rdar://103692353

Reviewed by Aditya Keerthi.

ButtonPart will handle drawing the button form control. This covers the parts'
types: Button, PushButton, DefaultButton and SquareButton. The following HTML
markup shows different buttons in the page:

<button>
<input type="button">
<input type="submit">
<input type="reset">
<input type="file">

On macOS, AppKit will be used to draw the platform controls through ButtonMac.

* Source/WebCore/Headers.cmake:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/controls/ButtonPart.h: Copied from Source/WebCore/platform/graphics/mac/controls/TextFieldMac.h.
* Source/WebCore/platform/graphics/controls/ControlFactory.h:
* Source/WebCore/platform/graphics/controls/PlatformControl.h:
(WebCore::PlatformControl::rectForBounds const):
* Source/WebCore/platform/graphics/mac/controls/ButtonMac.h: Copied from Source/WebCore/platform/graphics/mac/controls/TextFieldMac.h.
* Source/WebCore/platform/graphics/mac/controls/ButtonMac.mm: Added.
(WebCore::ButtonMac::ButtonMac):
(WebCore::ButtonMac::cellSize const):
(WebCore::ButtonMac::cellOutsets const):
(WebCore::ButtonMac::bezelStyle const):
(WebCore::ButtonMac::updateCellStates):
(WebCore::ButtonMac::rectForBounds const):
(WebCore::ButtonMac::draw):
* Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.h:
* Source/WebCore/platform/graphics/mac/controls/ControlFactoryMac.mm:
(WebCore::createButtonCell):
(WebCore::ControlFactoryMac::buttonCell const):
(WebCore::ControlFactoryMac::defaultButtonCell const):
(WebCore::ControlFactoryMac::createPlatformButton):
* Source/WebCore/platform/graphics/mac/controls/ControlMac.h:
* Source/WebCore/platform/graphics/mac/controls/ControlMac.mm:
(WebCore::ControlMac::setFocusRingClipRect):
(WebCore::ControlMac::inflatedRect):
* Source/WebCore/platform/graphics/mac/controls/TextFieldMac.h:
* Source/WebCore/platform/mac/ThemeMac.mm:
(WebCore::button):
(WebCore::ThemeMac::paint):
(WebCore::paintButton): Deleted.
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::createControlPart const):
* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::RenderThemeMac::canPaint const):
(WebCore::RenderThemeMac::canCreateControlPartForRenderer const):
* Source/WebKit/Shared/WebCoreArgumentCoders.cpp:
(IPC::ArgumentCoder<ControlPart>::decode):

Canonical link: https://commits.webkit.org/258492@main
